File: src/main/java/kr/syeyoung/dungeonsguide/dungeon/DungeonFacade.java
package kr.syeyoung.dungeonsguide.dungeon;

import com.google.common.util.concurrent.ThreadFactoryBuilder;
import kr.syeyoung.dungeonsguide.Main;
import kr.syeyoung.dungeonsguide.dungeon.pathfinding.CachedPathFinder;
import kr.syeyoung.dungeonsguide.dungeon.pathfinding.impl.ThetaStar;
import kr.syeyoung.dungeonsguide.dungeon.roomfinder.DungeonRoom;
import lombok.Setter;
import lombok.val;
import net.minecraftforge.common.MinecraftForge;
import org.joml.Vector3d;
import org.joml.Vector3i;

import java.io.IOException;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

public class DungeonFacade {


    public CachedPathFinder getCachedPathFinder() {
        return cachedPathFinder;
    }

    final CachedPathFinder cachedPathFinder;

    public final ExecutorService ex;

    public static DungeonFacade INSTANCE;

    public DungeonContext getContext() {
        return context;
    }

    @Setter
    private DungeonContext context;

    public DungeonFacade() {
        INSTANCE = this;
        ex = Executors.newCachedThreadPool(
                new ThreadFactoryBuilder().setNameFormat("Dg-AsyncPathFinder-%d").build()
        );
        cachedPathFinder = new CachedPathFinder();
    }

    public void init() {
        DungeonListener dgEventListener = new DungeonListener();
        MinecraftForge.EVENT_BUS.register(dgEventListener);
        try {
            DungeonRoomInfoRegistry.loadAll(Main.getConfigDir());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }


    public float calculatePathLenght(Vector3i from, Vector3i to, DungeonRoom r){
        float distance = -1;
        Vector3d fromv3 = new Vector3d(from.x, from.y, from.z);
        Vector3d tov3 = new Vector3d(to.x, to.y, to.z);
        val a = genPathfind(fromv3, tov3, r);
        List<Vector3d> b;
        try {
            b = a.get();
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }

        Vector3d last = null;
        for (Iterator<Vector3d> iterator = b.iterator(); iterator.hasNext(); ) {
            Vector3d vec3 = iterator.next();
            if(last != null){
                distance += vec3.distance(last);
            }
            last = vec3;
        }
        
        return distance;
    }


    Future<List<Vector3d>> genPathfind(Vector3d from, Vector3d to, DungeonRoom room){
        return ex.submit(() -> {
            ThetaStar pathFinder = new ThetaStar(room);
            pathFinder.pathfind(from, to,100);
            return pathFinder.getRoute();
        });
    }

}

File: src/main/java/kr/syeyoung/dungeonsguide/dungeon/DungeonFacade.kt
package kr.syeyoung.dungeonsguide.dungeon

import com.google.common.util.concurrent.ThreadFactoryBuilder
import kr.syeyoung.dungeonsguide.Main
import kr.syeyoung.dungeonsguide.dungeon.DungeonRoomInfoRegistry.loadAll
import kr.syeyoung.dungeonsguide.dungeon.pathfinding.CachedPathFinder
import kr.syeyoung.dungeonsguide.dungeon.pathfinding.impl.ThetaStar
import net.minecraftforge.common.MinecraftForge
import org.joml.Vector3d
import org.joml.Vector3i
import java.util.concurrent.ExecutionException
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.Future

class DungeonFacade {
    lateinit var cachedPathFinder: CachedPathFinder
    lateinit var ex: ExecutorService


    init {
        INSTANCE = this
    }

    fun init() {
        val dgEventListener = DungeonListener()
        MinecraftForge.EVENT_BUS.register(dgEventListener)
        loadAll(Main.getConfigDir())
        ex = Executors.newCachedThreadPool(
            ThreadFactoryBuilder().setNameFormat("Dg-AsyncPathFinder-%d").build()
        )
        cachedPathFinder = CachedPathFinder()
    }

    fun calculatePathLenght(from: Vector3i, to: Vector3i, r: DungeonRoom?): Float {
        var distance = -1f
        val fromv3 = Vector3d(from.x.toDouble(), from.y.toDouble(), from.z.toDouble())
        val tov3 = Vector3d(to.x.toDouble(), to.y.toDouble(), to.z.toDouble())
        val a = genPathfind(fromv3, tov3, r)
        val b: List<Vector3d>
        b = try {
            a.get()
        } catch (e: InterruptedException) {
            throw RuntimeException(e)
        } catch (e: ExecutionException) {
            throw RuntimeException(e)
        }
        var last: Vector3d? = null
        val iterator = b.iterator()
        while (iterator.hasNext()) {
            val vec3 = iterator.next()
            if (last != null) {
                distance += vec3.distance(last).toFloat()
            }
            last = vec3
        }
        return distance
    }

    fun genPathfind(from: Vector3d?, to: Vector3d?, room: DungeonRoom?): Future<List<Vector3d>> {
        return ex.submit<List<Vector3d>> {
            val pathFinder = ThetaStar(room!!)
            pathFinder.pathfind(from!!, to!!, 100f)
            pathFinder.route
        }
    }

    companion object {
        @JvmField
        var context: DungeonContext? = null
        lateinit var INSTANCE: DungeonFacade
    }
}